File: ai/src/main/java/org/pingel/causality/docalculus/DeleteObservation.java
package org.pingel.causality.docalculus;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.pingel.bayes.CausalModel;
import org.pingel.bayes.Probability;
import org.pingel.bayes.VariableNamer;
import org.pingel.forms.Variable;
import org.pingel.gestalt.core.Form;
import org.pingel.gestalt.core.Unifier;

public class DeleteObservation extends Rule {

    public List<Form> apply(Probability q, CausalModel m, VariableNamer namer) {

    		List<Form> results = new ArrayList<Form>();
        
        Set<Variable> Y = q.getQuestion();
        Set<Variable> X = q.getActions();
        CausalModel subModel = m.duplicate();
        subModel.getGraph().removeInputs(randomVariablesOf(X));
        
        for( Variable zObservation : q.getGiven() ) {
            Set<Variable> Z = new HashSet<Variable>();
            Z.add(zObservation);
            
            Set<Variable> W = new HashSet<Variable>();
            W.addAll(q.getGiven());
            W.remove(zObservation);

            Set<Variable> WX = new HashSet<Variable>();
            WX.addAll(W);
            WX.addAll(X);
            
            if( subModel.blocks(randomVariablesOf(q.getGiven()), randomVariablesOf(Z), randomVariablesOf(WX)) ) {
                Set<Variable> Ycopy = new HashSet<Variable>();
                Ycopy.addAll(Y);
                Set<Variable> Xcopy = new HashSet<Variable>();
                Xcopy.addAll(X);

                Probability probFactory = new Probability();
                Unifier unifier = new Unifier();
                unifier.put(probFactory.question, Ycopy);
                unifier.put(probFactory.given, W);
                unifier.put(probFactory.actions, Xcopy);
                Form f = probFactory.createForm(unifier);
                results.add(f);
            }
        }
        
        return results;
    }

}

File: ai/src/main/scala/org/pingel/causality/docalculus/DeleteObservation.scala
package org.pingel.causality.docalculus

import scala.collection._
import org.pingel.causality.CausalModel
import org.pingel.bayes.Probability
import org.pingel.bayes.VariableNamer
import org.pingel.forms.Variable
import org.pingel.gestalt.core.Form
import org.pingel.gestalt.core.Unifier

class DeleteObservation extends Rule {

  def apply(q: Probability, m: CausalModel, namer: VariableNamer) = {

    val results = new mutable.ListBuffer[Form]()

    val Y = q.getQuestion()
    val X = q.getActions()
    val subModel = m.duplicate()
    subModel.getGraph().removeInputs(randomVariablesOf(X))

    for (zObservation <- q.getGiven()) {

      val Z = Set[Variable]() + zObservation
      val W = Set[Variable]() ++ q.getGiven - zObservation
      val WX = Set[Variable]() ++ W ++ X

      if (subModel.blocks(randomVariablesOf(q.getGiven()), randomVariablesOf(Z), randomVariablesOf(WX))) {

        val Ycopy = Set[Variable]() ++ Y
        val Xcopy = Set[Variable]() ++ X

        val probFactory = new Probability()
        val unifier = new Unifier()
        unifier.put(probFactory.question, Ycopy)
        unifier.put(probFactory.given, W)
        unifier.put(probFactory.actions, Xcopy)
        results += probFactory.createForm(unifier)
      }
    }

    results.toList
  }

}
